cpp/oneapi/dal/table/backend/homogen_table_impl.cpp

```diff
@@ -91,206 +90,417 @@ bool has_array_data_kind(const Policy& policy, const array<Data>& array, const A
+    constexpr float type_ratio = static_cast<float>(sizeof(DataSrc)) / sizeof(DataDest);
+    const int64_t dst_avaliable_count = static_cast<int64_t>(src_remaining_count * type_ratio);
+    if (dst_avaliable_count < dst_count) {
+        throw dal::range_error("requested dst size bigger than available");
+    }
 
-    if (layout_ != data_layout::row_major) {
-        throw std::runtime_error("unsupported data layout"); // TODO: oneDAL exception
+    if (src.has_mutable_data()) {
+        // TODO: in future, when table knows about mutability of its data this branch shall be
+        // available only for builders, not for tables.
+        auto start_pointer = reinterpret_cast<DataDest*>(src.get_mutable_data() + src_start_index);
+        dst.reset(src, start_pointer, dst_count);
+    }
+    else {
+        auto start_pointer = reinterpret_cast<const DataDest*>(src.get_data() + src_start_index);
+        dst.reset(src, start_pointer, dst_count);
     }
+}
 
-    const auto table_dtype = meta_.get_data_type(0);
-    if (block_dtype == table_dtype && has_array_data_kind(policy, data_, kind)) {
-        if (data_.has_mutable_data()) {
-            auto row_data = reinterpret_cast<Data*>(data_.get_mutable_data());
-            auto row_start_pointer = row_data + rows.start_idx * col_count_;
-            block.reset(data_, row_start_pointer, range_size);
+class block_access_provider {
+private:
+    struct block_info {
+        block_info(int64_t row_count,
+                   int64_t column_count,
+                   int64_t row_offset,
+                   int64_t column_offset)
+                : row_count(row_count),
+                  column_count(column_count),
+                  row_offset(row_offset),
+                  column_offset(column_offset),
+                  element_count(row_count * column_count) {
+            detail::check_mul_overflow(row_count, column_count);
+
+            if (row_count <= 0 || column_count <= 0 || row_offset < 0 || column_offset < 0) {
+                throw dal::range_error("invalid parameters");
+            }
         }
-        else {
-            auto row_data = reinterpret_cast<const Data*>(data_.get_data());
-            auto row_start_pointer = row_data + rows.start_idx * col_count_;
-            block.reset(data_, row_start_pointer, range_size);
+
+        int64_t row_count;
+        int64_t column_count;
+        int64_t row_offset;
+        int64_t column_offset;
+        int64_t element_count;
+    };
+
+    struct origin_info {
+        origin_info(data_type dtype, int64_t row_count, int64_t column_count)
+                : dtype(dtype),
+                  row_count(row_count),
+                  column_count(column_count),
+                  element_count(row_count * column_count) {
+            detail::check_mul_overflow(row_count, column_count);
+
+            if (row_count <= 0 || column_count <= 0) {
+                throw dal::range_error("invalid parameters");
```

Comment:
Error message is useless for the user.

Recommendation: check as mush as possible preconditions in `block_access_provider` constructor or `pull_rows_impl`, etc., to report what's wrong with the argument. Put assertions here.